### ISSUE TITLE
dependsOnMethods not respected with parallel="methods"

### ISSUE THREAD

**Issue by kevinoid** [cbeust/testng#1379]
### TestNG Version

6.11

### Expected behavior

Tests are run only after the tests on which they depend (via `dependsOnMethods`) have finished, even when running tests in parallel.

### Actual behavior

When running in parallel and a test runner thread becomes free, it always runs another test, even if the only unfinished tests are ones where not all dependencies have completed.

### Is the issue reproductible on runner?

- [ ] Shell
- [X] Maven
- [X] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

(Likely reproducible on all.  I have only tested Maven and Gradle.)

### Test case sample

I have created a minimal test project at https://github.com/kevinoid/demo-testng-depends-parallel which demonstrates the issue.  The test class is:

```java
import org.testng.annotations.*;

public class DependsOnMethodsTest {
	private volatile boolean method1done = false;

	@Test
	public void method1() throws InterruptedException {
		System.out.println("Start method1");
		Thread.sleep(1000);
		method1done = true;
		System.out.println("Finish method1");
	}

	@Test
	public void method2() {
		System.out.println("Start method2");
		System.out.println("Finish method2");
	}

	@Test(dependsOnMethods = "method1")
	public void method3() {
		System.out.println("Start method3");
		assert method1done : "method1 has finished";
		System.out.println("Finish method3");
	}
}
```

testng.xml is:

```xml
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="Suite1" parallel="methods" preserve-order="false">
  <test name="Test1">
    <classes>
      <class name="DependsOnMethodsTest" />
    </classes>
  </test>
</suite>
```

When run the test fails due to `assert method1done`.  After a bit of tracing, the issue appears to be that when `method2` finishes, the call to `DynamicGraph#getFreeNodes()` from [`GraphThreadPoolExecutor#afterExecute()`](https://github.com/cbeust/testng/blob/6.11/src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java#L81) finds no free nodes then [ignores the dependencies](https://github.com/cbeust/testng/blob/6.11/src/main/java/org/testng/internal/DynamicGraph.java#L105).

It seems that, if no free nodes are available, but nodes are still running, `getFreeNodes` should return an empty list.  Subsequent calls can return nodes freed up on completion or ignore dependencies if required.  However, I wasn't confident enough in this analysis to submit a PR.

I'd also suggest logging a warning or error if the dependencies must be ignored to continue.  It seems that this would only result from circular or conflicting dependencies, right?

Thanks,
Kevin

**Comment by kevinoid** [comment]
I just came across #1287.  Looks like this issue is a duplicate of that one.  My mistake.  Closing in favor of that issue.
